File: ebook/ebook-by-claude/scripts/build-pdf.py
"""
전자책 PDF 빌드 스크립트 (Pandoc 방식)

Pandoc + XeLaTeX 또는 Pandoc → HTML 변환을 제공합니다.
기본(추천) 빌드는 Node.js 기반 build-pdf.js를 사용하세요.

사용법:
  python scripts/build-pdf.py                  # pandoc → HTML 생성 (PDF는 Node.js 빌드 추천)
  python scripts/build-pdf.py --engine=xelatex # XeLaTeX PDF (한글 폰트 필요)
"""

import os
import sys
import argparse
import subprocess
import re
from pathlib import Path
from datetime import datetime

# Windows 인코딩
if sys.stdout.encoding != "utf-8":
    sys.stdout.reconfigure(encoding="utf-8")
if sys.stderr.encoding != "utf-8":
    sys.stderr.reconfigure(encoding="utf-8")

# ─── 경로 설정 ────────────────────────────────────────────
PROJECT_ROOT = Path(__file__).resolve().parent.parent
MANUSCRIPT_DIR = PROJECT_ROOT / "manuscript"
IMAGES_DIR = PROJECT_ROOT / "images"
STYLES_DIR = PROJECT_ROOT / "styles"
OUTPUT_DIR = PROJECT_ROOT / "output"
CSS_FILE = STYLES_DIR / "ebook.css"

MANUSCRIPT_FILES = [
    "00-cover.md",
    "01-chapter1.md",
    "02-chapter2.md",
    "03-chapter3.md",
    "04-chapter4.md",
    "05-chapter5.md",
    "06-chapter6.md",
    "07-appendix.md",
    "08-closing.md",
]

CHAPTER_IMAGES = {
    "01-chapter1.md": "chapters/ch1-header.png",
    "02-chapter2.md": "chapters/ch2-header.png",
    "03-chapter3.md": "chapters/ch3-header.png",
    "04-chapter4.md": "chapters/ch4-header.png",
    "05-chapter5.md": "chapters/ch5-header.png",
    "06-chapter6.md": "chapters/ch6-header.png",
    "07-appendix.md": "chapters/ch7-header.png",
}

METADATA = {
    "title": "Claude로 전자책 작성하기",
    "author": "J.M",
    "date": datetime.now().strftime("%Y년 %m월"),
    "lang": "ko",
}

OUTPUT_FILENAME = "claude-ebook-guide-by-jm"


def merge_manuscripts() -> str:
    """원고 파일을 순서대로 병합하고 챕터 이미지를 삽입"""
    parts = []

    for filename in MANUSCRIPT_FILES:
        filepath = MANUSCRIPT_DIR / filename
        if not filepath.exists():
            print(f"  [WARN] {filename} 없음, 스킵")
            continue

        content = filepath.read_text(encoding="utf-8")
        content = re.sub(r'^---\n.*?\n---\n', '', content, flags=re.DOTALL)

        if filename in CHAPTER_IMAGES:
            img_path = IMAGES_DIR / CHAPTER_IMAGES[filename]
            if img_path.exists():
                rel_img = os.path.relpath(img_path, PROJECT_ROOT).replace("\\", "/")
                img_tag = f'\n![챕터 이미지]({rel_img}){{.chapter-header-image}}\n'
                content = re.sub(
                    r'(^#\s+.+$)', r'\1' + img_tag,
                    content, count=1, flags=re.MULTILINE,
                )

        parts.append(content.strip())

    return "\n\n---\n\n".join(parts)


def build_html():
    """Pandoc으로 HTML 생성 (CSS 인라인 포함)"""
    print("\n[BUILD] Pandoc -> HTML")
    print("=" * 50)

    print("  [1/3] 원고 병합 중...")
    merged = merge_manuscripts()
    merged_md = OUTPUT_DIR / "_merged.md"
    merged_md.write_text(merged, encoding="utf-8")
    print(f"        {len(merged):,}자 병합됨")

    print("  [2/3] Pandoc HTML 변환 중...")
    html_file = OUTPUT_DIR / f"{OUTPUT_FILENAME}-pandoc.html"

    pandoc_cmd = [
        "pandoc", str(merged_md),
        "-o", str(html_file),
        "--standalone",
        "--from=markdown+raw_html+fenced_divs+bracketed_spans",
        "--to=html5",
        f"--metadata=title:{METADATA['title']}",
        f"--metadata=author:{METADATA['author']}",
        f"--metadata=lang:{METADATA['lang']}",
        "--wrap=none",
        f"--resource-path={PROJECT_ROOT}",
    ]

    result = subprocess.run(pandoc_cmd, capture_output=True, text=True, encoding="utf-8")
    if result.returncode != 0:
        print(f"  [ERROR] Pandoc 실패:\n{result.stderr}")
        return False

    # CSS 인라인 삽입
    html_content = html_file.read_text(encoding="utf-8")
    css_content = CSS_FILE.read_text(encoding="utf-8")
    html_content = html_content.replace("</head>", f"\n<style>\n{css_content}\n</style>\n</head>")
    html_file.write_text(html_content, encoding="utf-8")

    print(f"  [3/3] 완료!")
    print(f"        출력: {html_file.name}")
    print(f"        이 HTML을 build-pdf.js (Puppeteer)로 PDF 변환하세요.")

    # 임시 MD 정리
    merged_md.unlink(missing_ok=True)
    return True


def build_xelatex():
    """Pandoc + XeLaTeX → PDF"""
    print("\n[BUILD] Pandoc + XeLaTeX")
    print("=" * 50)

    print("  [1/3] 원고 병합 중...")
    merged = merge_manuscripts()
    merged_md = OUTPUT_DIR / "_merged.md"
    merged_md.write_text(merged, encoding="utf-8")
    print(f"        {len(merged):,}자 병합됨")

    print("  [2/3] XeLaTeX PDF 변환 중...")
    pdf_file = OUTPUT_DIR / f"{OUTPUT_FILENAME}-xelatex.pdf"

    pandoc_cmd = [
        "pandoc", str(merged_md),
        "-o", str(pdf_file),
        "--pdf-engine=xelatex",
        "--from=markdown+raw_html+fenced_divs",
        f"--metadata=title:{METADATA['title']}",
        f"--metadata=author:{METADATA['author']}",
        f"--metadata=date:{METADATA['date']}",
        f"--metadata=lang:{METADATA['lang']}",
        "-V", "geometry:a4paper,top=25mm,bottom=25mm,left=22mm,right=22mm",
        "-V", "mainfont:Noto Sans KR",
        "-V", "monofont:D2Coding",
        "-V", "fontsize=11pt",
        "-V", "linestretch=1.7",
        "-V", "colorlinks=true",
        "-V", "linkcolor=blue",
        f"--resource-path={PROJECT_ROOT}",
    ]

    result = subprocess.run(pandoc_cmd, capture_output=True, text=True, encoding="utf-8")
    if result.returncode != 0:
        print(f"  [ERROR] XeLaTeX 실패:\n{result.stderr[:500]}")
        merged_md.unlink(missing_ok=True)
        return False

    size_mb = pdf_file.stat().st_size / (1024 * 1024)
    print(f"  [3/3] 완료!")
    print(f"        출력: {pdf_file.name} ({size_mb:.1f} MB)")

    merged_md.unlink(missing_ok=True)
    return True


def main():
    parser = argparse.ArgumentParser(description="전자책 PDF 빌드 (Pandoc)")
    parser.add_argument(
        "--engine", choices=["html", "xelatex"], default="html",
        help="빌드 방식 (기본: html)",
    )
    args = parser.parse_args()

    print("=" * 50)
    print("  전자책 PDF 빌드 (Pandoc)")
    print(f"  {METADATA['title']} -- {METADATA['author']}")
    print("=" * 50)

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    # pandoc 체크
    try:
        subprocess.run(["pandoc", "--version"], capture_output=True, check=True, timeout=10)
    except (FileNotFoundError, subprocess.CalledProcessError):
        print("  [ERROR] pandoc이 설치되어 있지 않습니다.")
        sys.exit(1)

    if args.engine == "xelatex":
        ok = build_xelatex()
    else:
        ok = build_html()

    if not ok:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: ebook/ebook-by-claude/scripts/build-pdf.js
/**
 * 전자책 PDF 빌드 스크립트 (Puppeteer)
 *
 * markdown-it → HTML → Puppeteer → PDF
 *
 * 사용법:
 *   node scripts/build-pdf.js              # 전체 빌드
 *   node scripts/build-pdf.js --preview    # 미리보기 (20페이지)
 *   node scripts/build-pdf.js --html-only  # HTML만 생성 (디버깅용)
 *   node scripts/build-pdf.js --keep-html  # HTML 파일 유지
 */

const fs = require('fs');
const path = require('path');
const puppeteer = require('puppeteer');
const MarkdownIt = require('markdown-it');

// ─── 경로 설정 ──────────────────────────────────────────
const PROJECT_ROOT = path.resolve(__dirname, '..');
const MANUSCRIPT_DIR = path.join(PROJECT_ROOT, 'manuscript');
const IMAGES_DIR = path.join(PROJECT_ROOT, 'images');
const STYLES_DIR = path.join(PROJECT_ROOT, 'styles');
const OUTPUT_DIR = path.join(PROJECT_ROOT, 'output');
const CSS_FILE = path.join(STYLES_DIR, 'ebook.css');

const OUTPUT_FILENAME = 'claude-ebook-guide-by-jm';

// 원고 파일 순서
const MANUSCRIPT_FILES = [
  '00-cover.md',
  '01-chapter1.md',
  '02-chapter2.md',
  '03-chapter3.md',
  '04-chapter4.md',
  '05-chapter5.md',
  '06-chapter6.md',
  '07-appendix.md',
  '08-closing.md',
];

// 챕터 → 헤더 이미지 매핑
const CHAPTER_IMAGES = {
  '01-chapter1.md': 'chapters/ch1-header.png',
  '02-chapter2.md': 'chapters/ch2-header.png',
  '03-chapter3.md': 'chapters/ch3-header.png',
  '04-chapter4.md': 'chapters/ch4-header.png',
  '05-chapter5.md': 'chapters/ch5-header.png',
  '06-chapter6.md': 'chapters/ch6-header.png',
  '07-appendix.md': 'chapters/ch7-header.png',
};

// 메타데이터
const METADATA = {
  title: 'Claude로 전자책 작성하기',
  author: 'J.M',
  lang: 'ko',
};

// ─── Markdown-it 설정 ───────────────────────────────────
const md = new MarkdownIt({
  html: true,
  linkify: true,
  typographer: true,
});

// ─── 유틸 함수 ──────────────────────────────────────────

/** 이미지 경로를 file:// URI로 변환 (Puppeteer 호환) */
function toFileUri(filePath) {
  const normalized = filePath.replace(/\\/g, '/');
  if (normalized.startsWith('/')) {
    return `file://${normalized}`;
  }
  return `file:///${normalized}`;
}

/** YAML 프론트매터 제거 */
function stripFrontmatter(content) {
  return content.replace(/^---\n[\s\S]*?\n---\n/, '');
}

/** 원고 파일을 읽고 병합 */
function mergeManuscripts() {
  const parts = [];

  for (const filename of MANUSCRIPT_FILES) {
    const filepath = path.join(MANUSCRIPT_DIR, filename);
    if (!fs.existsSync(filepath)) {
      console.log(`  [WARN] ${filename} 없음, 스킵`);
      continue;
    }

    let content = fs.readFileSync(filepath, 'utf-8');
    content = stripFrontmatter(content);

    // 이미지는 원고 마크다운에 이미 삽입되어 있으므로 추가 삽입 불필요
    // buildHtml()에서 상대경로 → 절대경로 변환 처리

    parts.push(content.trim());
  }

  return parts.join('\n\n');
}

/** Markdown → HTML 변환 후 완전한 HTML 문서 생성 */
function buildHtml(markdownContent) {
  let bodyHtml = md.render(markdownContent);
  const cssContent = fs.readFileSync(CSS_FILE, 'utf-8');

  // 모든 이미지 상대경로를 절대 file:// URI로 변환
  // ../images/... → file:///C:/Users/.../images/...
  const imagesAbsolute = toFileUri(IMAGES_DIR);
  bodyHtml = bodyHtml.replace(
    /src="\.\.\/images\//g,
    `src="${imagesAbsolute}/`
  );

  const html = `<!DOCTYPE html>
<html lang="${METADATA.lang}">
<head>
  <meta charset="utf-8">
  <meta name="author" content="${METADATA.author}">
  <title>${METADATA.title}</title>
  <style>
${cssContent}
  </style>
</head>
<body>
${bodyHtml}
</body>
</html>`;

  return html;
}

/** Puppeteer로 PDF 생성 */
async function generatePdf(htmlPath, pdfPath, options = {}) {
  const browser = await puppeteer.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox', '--font-render-hinting=none'],
  });

  const page = await browser.newPage();

  // HTML 파일 로드
  const fileUrl = toFileUri(htmlPath);
  await page.goto(fileUrl, { waitUntil: 'networkidle0', timeout: 60000 });

  // 폰트 로딩 대기
  await page.evaluateHandle('document.fonts.ready');
  await new Promise(r => setTimeout(r, 1000));

  // PDF 생성
  const pdfOptions = {
    path: pdfPath,
    format: 'A4',
    printBackground: true,
    margin: {
      top: '25mm',
      bottom: '25mm',
      left: '22mm',
      right: '22mm',
    },
    displayHeaderFooter: true,
    headerTemplate: `
      <div style="width:100%; font-size:8pt; font-family:'Noto Sans KR',sans-serif; color:#94A3B8; padding:0 22mm;">
        <span style="float:left;">${METADATA.title}</span>
        <span style="float:right;">${METADATA.author}</span>
      </div>`,
    footerTemplate: `
      <div style="width:100%; text-align:center; font-size:9pt; font-family:'Noto Sans KR',sans-serif; color:#94A3B8;">
        <span class="pageNumber"></span>
      </div>`,
    preferCSSPageSize: false,
  };

  // 미리보기 모드: 페이지 제한
  if (options.previewPages) {
    pdfOptions.pageRanges = `1-${options.previewPages}`;
  }

  await page.pdf(pdfOptions);
  await browser.close();
}

// ─── 메인 ───────────────────────────────────────────────
async function main() {
  const args = process.argv.slice(2);
  const isPreview = args.includes('--preview');
  const htmlOnly = args.includes('--html-only');
  const keepHtml = args.includes('--keep-html');

  console.log('='.repeat(55));
  console.log(`  전자책 PDF 빌드 (Puppeteer)`);
  console.log(`  ${METADATA.title} — ${METADATA.author}`);
  console.log('='.repeat(55));
  console.log();

  // 출력 디렉토리 생성
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });

  // 1. 원고 병합
  console.log('[1/3] 원고 병합 중...');
  const merged = mergeManuscripts();
  const charCount = merged.length.toLocaleString();
  console.log(`      완료: ${charCount}자, ${MANUSCRIPT_FILES.length}개 파일`);

  // 2. HTML 변환
  console.log('[2/3] HTML 변환 중...');
  const html = buildHtml(merged);
  const htmlPath = path.join(OUTPUT_DIR, `${OUTPUT_FILENAME}.html`);
  fs.writeFileSync(htmlPath, html, 'utf-8');
  console.log(`      HTML 저장: ${path.basename(htmlPath)}`);

  if (htmlOnly) {
    console.log('\n[완료] --html-only 모드. HTML만 생성했습니다.');
    console.log(`      ${htmlPath}`);
    return;
  }

  // 3. PDF 생성
  const pdfFilename = isPreview
    ? `${OUTPUT_FILENAME}-preview.pdf`
    : `${OUTPUT_FILENAME}.pdf`;
  const pdfPath = path.join(OUTPUT_DIR, pdfFilename);

  console.log(`[3/3] PDF 생성 중${isPreview ? ' (미리보기 20p)' : ''}...`);

  try {
    await generatePdf(htmlPath, pdfPath, {
      previewPages: isPreview ? 20 : null,
    });

    const stats = fs.statSync(pdfPath);
    const sizeMb = (stats.size / (1024 * 1024)).toFixed(1);
    console.log(`      PDF 저장: ${pdfFilename} (${sizeMb} MB)`);
  } catch (err) {
    console.error(`  [ERROR] PDF 생성 실패: ${err.message}`);
    process.exit(1);
  }

  // HTML 정리
  if (!keepHtml) {
    fs.unlinkSync(htmlPath);
  }

  // 결과 요약
  console.log();
  console.log('='.repeat(55));
  console.log('  빌드 완료!');
  console.log(`  출력: output/${pdfFilename}`);
  console.log('='.repeat(55));
}

main().catch(err => {
  console.error('빌드 오류:', err);
  process.exit(1);
});
